wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/convert/tool/kubernetes/KubernetesToolToWorkspaceApplier.java

```diff
@@ -125,7 +125,7 @@ private String retrieveContent(
     if (isNullOrEmpty(recipeFileContent)) {
       throw new DevfileException(
           format(
-              "The local file '%s' defined in tool '%s' is empty.",
+              "The local file '%s' defined in tool '%s' is empty or could not be retrieved.",
```

Comment:
As far as I see if there is any exception during content retrieving than an exception should be thrown here https://github.com/eclipse/che/pull/12800/files#diff-433beefd65ee6f2da3b5fc799536bd3bR120. 
Please elaborate more on why you did this change.

Reply:
Would `or could not be found` work better? I can understand this check for empty contents, but I don't really understand it for `null`. Hence, I thought that a `null` file content really means some kind of error condition not captured by `IOException`.

Reply:
The semantics of `null` content is really not described anywhere actually and depends on the free form `Function<String, String>` supplied to the `URLFactoryBuilder#createFactoryFromDevfile`. That's why I made this change here and updated the javadoc of the `createFactoryFromDevfile` to allow returning `null` on failure to map the filename to a URL.

Reply:
It's not really `Function<String, String>` but the interface. See https://github.com/eclipse/che/blob/master/wsmaster/che-core-api-devfile/src/main/java/org/eclipse/che/api/devfile/server/FileContentProvider.java#L36

> but I don't really understand it for `null`

I think checking on null can be removed here. 
Or leave it to have code more safer, but it should not be null here.

Reply:
The only non-test implementation of that interface is the `URLFileContentProvider` in this PR or the lambda in `URLFactoryBuilder#createFactoryFromDevfile` in master. That's why I said that it depends on the `Function<String, String>`.

Frankly the whole flow is overcomplicated for what it achieves, but I wanted to make the minimal changes necessary.

If `null` is not valid, we should remove the checks for it, IMHO. WDYT?

Reply:
If implementation will not follow interface then NPE will be thrown.
I would leave checking on null, maybe with dedicated message for null content

Reply:
I've reworked the implementation on top of the latest master which already improved on the situation, making this impl simpler.